core/templates/pages/exploration-player-page/services/stats-reporting.service.spec.ts

```diff
@@ -0,0 +1,292 @@
+    statsReportingBackendApiService = TestBed.inject(
+      StatsReportingBackendApiService);
+    messengerService = TestBed.inject(MessengerService);
+    siteAnalyticsService = TestBed.inject(SiteAnalyticsService);
+    playthroughService = TestBed.inject(PlaythroughService);
+    urlService = TestBed.inject(UrlService);
+  });
+
+  beforeEach(() => {
+    recordExplorationStartedSpy = spyOn(
+      statsReportingBackendApiService, 'recordExpStartedAsync')
+      .and.returnValue(Promise.resolve({}));
+    postStatsSpy = spyOn(statsReportingBackendApiService, 'postsStatsAsync')
+      .and.returnValue(Promise.resolve({}));
+    recordExplorationActuallyStartedSpy = spyOn(
+      statsReportingBackendApiService, 'recordExplorationActuallyStartedAsync')
+      .and.returnValue(Promise.resolve({}));
+    recordSolutionHitSpy = spyOn(
+      statsReportingBackendApiService, 'recordSolutionHitAsync')
+      .and.returnValue(Promise.resolve({}));
+    recordLeaveForRefresherExpSpy = spyOn(
+      statsReportingBackendApiService, 'recordLeaveForRefresherExpAsync')
+      .and.returnValue(Promise.resolve({}));
+    recordStateHitSpy = spyOn(
+      statsReportingBackendApiService, 'recordStateHitAsync')
+      .and.returnValue(Promise.resolve({}));
+    recordStateCompletedSpy = spyOn(
+      statsReportingBackendApiService, 'recordStateCompletedAsync')
+      .and.returnValue(Promise.resolve({}));
+    recordExplorationCompletedSpy = spyOn(
+      statsReportingBackendApiService, 'recordExplorationCompletedAsync')
+      .and.returnValue(Promise.resolve({}));
+    recordAnswerSubmittedSpy = spyOn(
+      statsReportingBackendApiService, 'recordAnswerSubmittedAsync')
+      .and.returnValue(Promise.resolve({}));
+    recordMaybeLeaveEventSpy = spyOn(
+      statsReportingBackendApiService, 'recordMaybeLeaveEventAsync')
+      .and.returnValue(Promise.resolve({}));
+    recordAnswerSubmitActionSpy = spyOn(
+      playthroughService, 'recordAnswerSubmitAction')
+      .and.returnValue(null);
+    spyOn(messengerService, 'sendMessage').and.returnValue(null);
+    spyOn(siteAnalyticsService, 'registerNewCard').and.returnValue(null);
+    spyOn(siteAnalyticsService, 'registerStartExploration')
+      .and.returnValue(null);
+    spyOn(siteAnalyticsService, 'registerFinishExploration')
+      .and.returnValue(null);
+    spyOn(siteAnalyticsService, 'registerCuratedLessonCompleted')
+      .and.returnValue(null);
+    spyOn(playthroughService, 'recordExplorationStartAction')
+      .and.returnValue(null);
+    spyOn(playthroughService, 'recordExplorationQuitAction')
+      .and.returnValue(null);
+    spyOn(playthroughService, 'storePlaythrough')
+      .and.returnValue(null);
+    spyOn(contextService, 'isInExplorationEditorPage').and.returnValue(true);
+    spyOn(contextService, 'isInQuestionPlayerMode').and.returnValue(true);
+    spyOn(urlService, 'getUrlParams')
+      .and.returnValue({classroom_url_fragment: 'classroom'});
```

Comment:
Why are we creating so many spyOn in beforeEach block? It should be created in `it` block and only the required spyOn, right?

Reply:
Most of the test cases use multiple spies.
In order to decrease the redundancy of code I've placed them in `beforeEach` block, that way it is easier to read code too.
I think both methods are identical.

Reply:
I think to make each test case readable we should write spies in `it` function where we are testing behavior based on the spied function!

Reply:
done!